nixos/modules/services/networking/wireguard.nix

```diff
@@ -420,20 +430,36 @@ in
       ) cfg.interfaces);
   in {
 
-    assertions = (attrValues (
-        mapAttrs (name: value: {
-          assertion = (value.privateKey != null) != (value.privateKeyFile != null);
-          message = "Either networking.wireguard.interfaces.${name}.privateKey or networking.wireguard.interfaces.${name}.privateKeyFile must be set.";
-        }) cfg.interfaces))
-      ++ (attrValues (
-        mapAttrs (name: value: {
-          assertion = value.generatePrivateKeyFile -> (value.privateKey == null);
-          message = "networking.wireguard.interfaces.${name}.generatePrivateKeyFile must not be set if networking.wireguard.interfaces.${name}.privateKey is set.";
-        }) cfg.interfaces))
-        ++ map ({ interfaceName, peer, ... }: {
-          assertion = (peer.presharedKey == null) || (peer.presharedKeyFile == null);
-          message = "networking.wireguard.interfaces.${interfaceName} peer «${peer.publicKey}» has both presharedKey and presharedKeyFile set, but only one can be used.";
-        }) all_peers;
+    assertions = builtins.concatLists (attrValues (
+      mapAttrs (name: value:
+        let
+          interfaceAssertions = [
+            {
+              assertion = (value.privateKey != null) != (value.privateKeyFile != null);
+              message = "Either networking.wireguard.interfaces.${name}.privateKey or networking.wireguard.interfaces.${name}.privateKeyFile must be set.";
+            }
+            {
+              assertion = value.generatePrivateKeyFile -> (value.privateKey == null);
+              message = "networking.wireguard.interfaces.${name}.generatePrivateKeyFile must not be set if networking.wireguard.interfaces.${name}.privateKey is set.";
+            }
+          ];
+          peerAssertions = builtins.concatLists (map (peer:
+            let
+              peerName = if peer.publicKey != null then peer.publicKey else peer.publicKeyFile;
+            in
+            [
+              {
```

Comment:
I think we need an assert that one of: publicKey or publicKeyFile is set? I don't see one, but it is possible.

Reply:
That case would be caught by this assertion as well since `(null != null) != (null != null)` is false. I basically copied the pre-existing logic for `privateKey` and `privateKeyFile`.